aten/src/ATen/native/cuda/StochasticRounding.cu

```diff
@@ -0,0 +1,62 @@
+#include <ATen/ATen.h>
+#include <ATen/native/cuda/stochastic_rounding.cuh>
+
+
+namespace at {
+namespace native {
+
+template <typename input_t, typename output_t>
+__global__ void stochastic_rounding_kernel(
+    const input_t* input,
+    output_t* output,
+    const int64_t numel,
+    std::pair<uint64_t, uint64_t> seed_and_offset) {
+  int tid = blockIdx.x * blockDim.x + threadIdx.x;
+  curandStatePhilox4_32_10_t state;
+  curand_init(seed_and_offset.first, tid, seed_and_offset.second, &state);
+
+  for (int64_t i = tid; i < numel; i += blockDim.x * gridDim.x) {
+    float inp = static_cast<float>(input[i]);
+    output[i] = round_stochastically<output_t>(inp, curand_uniform(&state));
+  }
+}
+
+Tensor stochastic_rounding_cuda(const Tensor& input, c10::optional<Generator> gen_) {
+
+  TORCH_CHECK(input.is_contiguous());
+
+  if (input.scalar_type() == kHalf) {
+    return input;
+  }
+
+  Tensor output = at::empty_like(input, input.options().dtype(kHalf), input.suggest_memory_format());
+  const int64_t numel = input.numel();
+  if (numel == 0) {
+    return output;
+  }
+
+  const int block = 256;
+  const int blocks_per_sm = at::cuda::getCurrentDeviceProperties()->maxThreadsPerMultiProcessor / block;
```

Comment:
This is only correct if the kernel's number of registers per thread is <= 32, otherwise register pressure will limit your occupancy.  You can recompile your kernels with `-ptxas-options=-v` as an nvcc option to doublecheck how many registers they use (this is easiest to do with the kernels in an extension, I'm not sure how you would pass it to pytorch).